### PR TITLE
LCS on Permutations

### PR DESCRIPTION
**Pull request by dongliu0426**

### CODE REVIEW

File: content/4_Gold/LIS.mdx [hunk 1]
```diff
@@ -338,6 +338,62 @@ class PCB{
...
+
+### $\mathcal O(N^2)$
+
+See [here](/gold/paths-grids#solution---longest-common-subsequence).
+
+### $\mathcal O(N\log N)$
+
+In the $\mathcal O(N^2)$ solution, we increase the DP arrray only if $a[i] = b[j]$, but since both arrays are permutations of length $N$, for each $a[i]$, there must be only one matching element in $b$.  
+
+Thus we can reorder the elements in $b$ such that $b[i]$ becomes $\texttt{pos}[b[i]]$ where $\texttt{pos}[i]$ represents the index of $i$ in $a$. After reordering $b$, our answer would be the LIS of $b$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:353]: More intuitive if we create array $c[i]$ where $c[i] = a^-1[b[i]]$

Additionally, possibly mention that the final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$

**Review by frodakcin — CHANGES_REQUESTED** [review]

File: content/4_Gold/LIS.mdx (hunk 1 above)
Comment (frodakcin) [reply content/4_Gold/LIS.mdx:353]: in other words, every increasing subsequence $x_{1 \dots k}$ in $c$ corresponds to a common subsequence between $a$ and $b$

Comment (dongliu0426) [reply content/4_Gold/LIS.mdx:357]: > More intuitive if we create array $c[i]$ where $c[i] = a^-1[b[i]]$
> 
> Additionally, possibly mention that the final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$

what does $c[i] = a^-1[b[i]]$ mean?

Comment (bqi343) [reply content/4_Gold/LIS.mdx:357]: a^{-1}[i] = pos[i] :/

Comment (frodakcin) [reply content/4_Gold/LIS.mdx:361]: Oops yea that's what I meant. Basically store the "reordered" b array into a new array because it means something different than just reordering it.

File: content/4_Gold/LIS.mdx [hunk 2]
```diff
@@ -338,6 +342,68 @@ class PCB{
...
+
+### Efficient Solution
+
+**Time Complexity**: $\mathcal O(N\log N)$
+
+In the $\mathcal O(N^2)$ solution, we increase the DP arrray only if $a[i] = b[j]$, but since both arrays are permutations of length $N$, for each $a[i]$, there must be only one matching element in $b$.  
+
+Let us create an array $c$ where $c[i]$ is $a^{-1}[b[i]]$. Our final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$.
+
+In other words, every increasing subsequence $x_{1 \dots k}$ in $c$ corresponds to a common subsequence between $a$ and $b$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:363]:
```suggestion
Let us create an array $c$ where $c[i]$ is $a^{-1}[b[i]]$ ($a^{-1}$ is the inverse of $a$: $a^{-1}[x]$ is the index where element $x$ is located. Specifically $a[a^{-1}[x]]=x and a^{-1}[a[i]]=i$). 

Notice that every increasing subsequence $x_{1 \dots k}$ in $c$ corresponds to a common subsequence between $a$ and $b$. Increasing subsequence $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the the common subsequence $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$. Thus, the length of the longest common subsequence between $a$ and $b$ is the longest increasing subsequence of $c$.
```

Something like this probably flows better (my initial suggestion was more of some stuff that would be good to have as opposed to structured info). Also, I notice there's leading spaces - I didn't add them but format it however you like.

I made it a bit wordy, maybe the thing explaining a^-1 is a bit messy/unclear... modify it how you like

File: content/4_Gold/LIS.mdx [hunk 3]
```diff
@@ -338,6 +342,68 @@ class PCB{
...
+int n, a[N], pos[N], b[N];
+vector<int> lis;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+
+	cin >> n;
+	for (int i = 1; i <= n; ++i) {
+		cin >> a[i];
+		pos[a[i]] = i;
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:385]: some sort of comment, maybe "pos is the inverse of permutation a" or "pos is a^-1 as describe above"

File: content/4_Gold/LIS.mdx [hunk 4]
```diff
@@ -338,6 +342,68 @@ class PCB{
...
+		if (p == lis.size())
+			lis.push_back(b[i]);
+		else
+			lis[p] = b[i];
+	}
+	cout << lis.size() << '\n';
+}
+```
+
+</CPPSection>
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:402]: Would be nice if you explicitly created array c in the code as well. If not, add a comment saying smth along the lines of "instead of creating permutation $c$, we modify permutation b"

**Review by frodakcin — CHANGES_REQUESTED** [review]

File: content/4_Gold/LIS.mdx (hunk 2 above)
Comment (frodakcin) [reply content/4_Gold/LIS.mdx:363]: Or you could just define pos the way you did before instead of using the a^-1 notation

File: content/4_Gold/LIS.mdx [hunk 5]
```diff
@@ -338,6 +342,92 @@ class PCB{
...
+
+See [here](/gold/paths-grids#solution---longest-common-subsequence).
+
+### Efficient Solution
+
+**Time Complexity**: $\mathcal O(N\log N)$
+
+In the $\mathcal O(N^2)$ solution, we increase the DP arrray only if $a[i] = b[j]$, but since both arrays are permutations of length $N$, for each $a[i]$, there must be only one matching element in $b$.  
+
+Let us create an array $pos$ where $pos[x]$ is the index of $x$ in $a$ ($a[pos[x]] = x$). Then, we can create another array, $c$ where $c[i]$ stores $pos[b[i]$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:381]:
```suggestion
Let us create an array $pos$ where $pos[x]$ is the index of $x$ in $a$ ($a[pos[x]] = x$). Then, we can create another array, $c$ where $c[i]$ stores $pos[b[i]]$.
```

**Review by frodakcin — CHANGES_REQUESTED** [review]

**Review by frodakcin — APPROVED** [review]
Is there an existing problem to submit to?

**Comment by dongliu0426** [comment]
> Is there an existing problem to submit to?

erm its on [Luogu](https://www.luogu.com.cn/problem/P1439) (which is in chinese) 

could you add it to the cf gym or smth?

**Comment by frodakcin** [comment]
Ok, I'll merge in a couple hours if nothing else happens.
